### DISCUSSION TITLE
ONVIF device test tool unexpected passes for unauthenticated user

### DISCUSSION THREAD

**Discussion by bigbrobrody** [onvif/testspecs#31] [Q&A] [answered]
We have executed two ONVIF device test tool tests as an unauthenticated user and would expect the tests to fail.

The tests are:
- SEARCH-2-1-10-v14.12 GET RECORDING SEARCH RESULTS WITH INVALID SEARCHTOKEN
- SEARCH-2-1-11-v14.12 END SEARCH WITH INVALID SEARCHTOKEN

The test tool is correctly identifying a discrepancy in the response at step 3 of each test, but identifies the test step as passed and the overall test as passed. Shouldn't the result be reported as a failure?

I've reproduced below the output from the device test tool.

SEARCH-2-1-10-v14.12 GET RECORDING SEARCH RESULTS WITH INVALID SEARCHTOKEN

STEP 1 - Get Search Service address
StepStart: 2022-03-24T17:01:49.5076932Z
Transmit done
Receive done
STEP PASSED

STEP 2 - Connect to Search service
StepStart: 2022-03-24T17:01:49.6026921Z
Search service address:  http://[redacted]/Services/Search/SearchService.svc
STEP PASSED

STEP 3 - Get recordings search result with invalid token (c0812816)
StepStart: 2022-03-24T17:01:49.6276914Z
Transmit done
Receive done
SOAP fault returned
Code: Sender
Subcode: NotAuthorized
Reason: Sender not Authorized
WARNING: The SOAP FAULT returned from the DUT is not as expected: fault subcodes sequence is incorrect. Expected: env:Sender/ter:InvalidArgVal/ter:InvalidToken, actual: env:Sender/ter:NotAuthorized
STEP PASSED

TEST PASSED
SEARCH-2-1-11-v14.12 END SEARCH WITH INVALID SEARCHTOKEN

STEP 1 - Get Search Service address
StepStart: 2022-03-24T17:01:49.6676948Z
Transmit done
Receive done
STEP PASSED

STEP 2 - Connect to Search service
StepStart: 2022-03-24T17:01:49.7826884Z
Search service address:  http://[redacted]/Services/Search/SearchService.svc
STEP PASSED

STEP 3 - End search with invalid token (b85286f5)
StepStart: 2022-03-24T17:01:49.7926971Z
Transmit done
Receive done
SOAP fault returned
Code: Sender
Subcode: NotAuthorized
Reason: Sender not Authorized
WARNING: The SOAP FAULT returned from the DUT is not as expected: fault subcodes sequence is incorrect. Expected: env:Sender/ter:InvalidArgVal/ter:InvalidToken, actual: env:Sender/ter:NotAuthorized
STEP PASSED

TEST PASSED

**Comment by baldvin137** [comment, marked as answer]
Excellent question. 

The answer can be found in the [Recording Search Test Specification](https://www.onvif.org/wp-content/uploads/2021/01/[ONVIF_Recording_Search_Device_Test_Specification_20.06.pdf). Both these test cases have the following note: _"Other faults than specified in the test are acceptable, though the specified are preferable."_

The reason why other faults than those specified are acceptable, is that specific error handling is not mandatory in the ONVIF Network Interface Specifications (although it is recommended). It is debatable whether this is good or bad, but this is the reason why there are very few tests in the ONVIF Device Test Tool that test error handling and why test cases that do, tend to be very forgiving (like these two you mention).

If you are interested in testing the authentication behavior of a device, then (instead of executing tests as an unauthenticated or unauthorized user) you can use the Access Policy test case. It is not a part of the Device Test Tool Conformance mode (because authentication levels are not mandatory) but it is available in Diagnostic mode. Information about this test can be found in the [Access Policy Device Test Specification](https://www.onvif.org/wp-content/uploads/2021/12/ONVIF_Access_Policy_Device_Test_Specification_21.12.pdf). Please note that as this test is in Diagnostic mode only it has not been tested as rigourously as tests in Conformance mode. If you have any feedback on the Access Policy test case it would be much appreciated.

Hope this helps!
